BXMSDOC-5869: Add test scenarios content for Kogito.

sterobin opened the pull request:
@yesamer , can you review and verify the new test scenarios doc content I've added for Kogito? Links below. I created the test scenarios doc as usual, but also add a section on test scenarios to our Kogito getting started doc/chapter.

@dupliaka , would you like to review as well? Not required, but I'd love your review and feedback!

@karreiro , keeping you in the loop as you requested. Your review is also welcome!

I started with the [BA 7.x content](https://access.redhat.com/documentation/en-us/red_hat_process_automation_manager/7.7/html-single/testing_a_decision_service_using_test_scenarios/index) as a base, but greatly simplified and streamlined for what I would like to be an easier user experience with test scenarios. I also excluded non-applicable content, expanded where things were new, and adapted to the Kogito doc standard for consistency.

I've pinged you with some preliminary comments or questions in the PR. Thanks!

Links:
* [JIRA](https://issues.redhat.com/browse/BXMSDOC-5869)
* Enterprise doc previews:
** [Testing decision services](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5869_DS/#testing_decision_services_using_test_scenarios_in_kogito)
** [Testing the decision logic](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5869_CR/#proc-kogito-testing-decision-logic_kogito-creating-running) (getting started doc) - Abbreviated version. I wanted to include this here as well for the user's getting started experience. It conforms to the flow and format of that tutorial.
** [Test scenarios release note](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5869_RN/#kogito_tooling) - I went ahead and just added this while I was at it, to save you the time @yesamer 
* Community doc previews:
** [Testing decision services](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5869_COMM/#_testing_decision_services_using_test_scenarios_in_kogito) - Same as enterprise, but conditioned
** [Testing the decision logic](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5869_COMM/#proc-kogito-testing-decision-logic_kogito-creating-running) - Ditto
** [Test scenarios release note](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5869_COMM/#_kogito_tooling) - Ditto

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/test-scenarios/proc-test-scenarios-collections.adoc:
```diff
@@ -0,0 +1,45 @@
+[id='proc-test-scenarios-collections_{context}']
+= Defining list values in test scenarios
```
@yesamer , I refined this a bit to focus on `list` specifically, rather than both `list` and `map`, because it seems like `map` currently doesn't apply. If it does apply, can you clarify how a user would configure that for DMN data types and test scenarios? I only see how to do it as a `list`.

yesamer replied:
@sterobin Agree with you choice. `Map`s are used in **Rule** based test scenarios, which is currently not supported. As the previous comment, a description of `Map` field can be included in the moment we will support Rule based test scenarios.

Regarding the bug on `List` collection we currently have: most probably it can't be available for version 0.4.2, because PR is not merged yet. How we can manage it?

sterobin replied:
@yesamer, ack on the `List` support delay beyond 0.4.2. I've commented out that section in the source but have left the doc preview for review purposes, so that when it becomes available, it has already been reviewed and approved here and I can re-include it once available. Thanks.

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/test-scenarios/ref-test-scenarios-supported-feel-data-types.adoc:
```diff
@@ -0,0 +1,35 @@
+[id='ref-test-scenarios-supported-feel-data-types_{context}']
+= Supported FEEL data types in DMN test scenarios
```
@yesamer , I tweaked this a bit to be more consistent with what we describe elsewhere in DMN documentation, and with the cheatsheet in the UI. Let me know if anything's off of unclear.

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/test-scenarios/proc-test-scenarios-creating.adoc:
```diff
@@ -0,0 +1,259 @@
+[id='proc-test-scenarios-creating_{context}']
+= Creating and running test scenarios in {PRODUCT}
+
+You can use the {PRODUCT} test scenario modeler in VSCode to design test scenarios for testing the functionality of business decisions and business decision data during project development.
```
@yesamer , @dupliaka , just FYI that I decided to go with "test scenario modeler" for consistency with the other Kogito modelers, such as DMN modeler, BPMN modeler, etc. Plus I understand that the scesim modeler will be added to the Business Modeler Hub at some point. Let me know if you have any objections. I've referred to it as such throughout.

yesamer commented:
@sterobin Thank you Stetson, I'll give you a feedback. @danielezonca can you please review it?

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/test-scenarios/con-test-scenarios.adoc:
```diff
@@ -0,0 +1,15 @@
+.Example test scenarios for `PersonDecisions` DMN decision logic
+image::kogito/creating-running/kogito-test-scenario-example-person.png[Image of PersonDecisionsTest test scenario]
+
+When you run the test scenarios in your {PRODUCT} project, all test scenarios in all `.scesim` files are run as part of the same command. However, individual test scenarios are independent, so one scenario cannot affect or modify another. You can run test scenarios at any time during project development. You do not need to compile or deploy your decision service to run test scenarios, but your {P ...
```
Done. I added the following text to the end of the procedure, in the next file. Good with you?

>To update and re-run a test scenario, ensure that you save the updated `.scesim` file before you run the new scenarios. If you do not save, the {PRODUCT} runtime runs the previously saved version of the test scenario.

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/test-scenarios/con-test-scenarios.adoc:
```diff
@@ -0,0 +1,15 @@
+
+When you run the test scenarios in your {PRODUCT} project, all test scenarios in all `.scesim` files are run as part of the same command. However, individual test scenarios are independent, so one scenario cannot affect or modify another. You can run test scenarios at any time during project development. You do not need to compile or deploy your decision service to run test scenarios, but your {P ...
+
+All data objects related to a test scenario must be in the same project package as the test scenario. Assets in the same package are imported by default.
```
Done. Commented out for now.

sterobin commented:
@yesamer, ack on the `List` support delay beyond 0.4.2. I've commented out that section in the source but have left the doc preview for review purposes, so that when it becomes available, it has already been reviewed and approved here and I can re-include it once available. Thanks.

danielezonca commented on doc-content/kogito-docs/src/main/asciidoc/test-scenarios/proc-test-scenarios-creating.adoc:
```diff
@@ -0,0 +1,259 @@
+
+.Prerequisites
+* https://code.visualstudio.com/[VSCode] 1.44.0 or later is installed.
+* The {PRODUCT} https://github.com/kiegroup/kogito-tooling/releases[VSCode extension] is installed and enabled in your VSCode IDE. For information about enabling the VSCode extension, see
```
I think we should mention `VSCode extension 0.4.2 or later` like we are doing for VSCode version.
Should we want to mention that Test Scenario runner is compatible only with JUnit5 vintage? If you plan to use JUnit 4 in your project it doesn't work

sterobin replied:
@danielezonca , good points. I've updated to say `VSCode extension 0.4.2 or later`, and I also updated the JUnit mention to specify and link to [JUnit 5](https://junit.org/junit5/), per our gchat conversation. Thanks!

karreiro commented on doc-content/kogito-docs/src/main/asciidoc/test-scenarios/ref-test-scenarios-supported-feel-data-types.adoc:
```diff
@@ -0,0 +1,35 @@
+|Boolean values
+|`true`, `false`, and `null`.
+
+|Dates, time, and date and time
```
I'm not sure if I'm missing something, but I do think that `.scesim` supports duration as well. Since users can set it on decision/input nodes:

<img width="378" alt="Screen Shot 2020-05-08 at 14 24 55" src="https://user-images.githubusercontent.com/1079279/81431593-c4405280-9137-11ea-82fd-8550bef1a9ec.png">

So, we may add `duration("P4DT3H2M1S")` and `duration("P1Y2M")`.

wdyt @yesamer @sterobin ?

sterobin replied:
@yesamer @karreiro , my interpretation is that the test scenarios supported a subset of the supported DMN data types in the DMN modeler. Are you saying that all data types supported in the DMN modeler are supported no prob in the test scenario modeler? Meaning, whatever data types you use in the DMN modeler are all perfectly sustained in test scenarios? If this is the case, then imo, we don't even need this section at all since it doesn't provide much value. Not even sure in that case of the purpose of the list of supported data types in the scenario cheatsheet. Maybe I've misunderstood.

sterobin replied:
@yesamer @karreiro Updated the table per our gchat convo. Can view in the same doc preview links above, now updated.

dupliaka commented on doc-content/kogito-docs/src/main/asciidoc/creating-running/proc-kogito-testing-decision-logic.adoc:
```diff
@@ -0,0 +1,477 @@
+    </scesimData>
+  </background>
+  <settings>
+    <dmnFilePath>src/main/resources/PersonDecisions.dmn</dmnFilePath>
```
You wrote that scesim file should be under
`src/main/test/
          └── java/testscenario
          └── resources
but path here is not the same (src/main/resources)
could you please change it?

sterobin replied:
@dupliaka , this is actually correct because it's referring to the DMN file (not the scesim file), which is indeed in `src/main/resources`.

dupliaka replied:
Oh, indeed sorry

dupliaka commented:
FYI
Dead links not related to the PR in all community docs for example:
![image](https://user-images.githubusercontent.com/21653197/81565733-213b4300-939a-11ea-8acf-29f9947275a5.png) in
[Testing decision services](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5869_COMM/#_testing_decision_services_using_test_scenarios_in_kogito)

yesamer commented on doc-content/kogito-docs/src/main/asciidoc/test-scenarios/proc-test-scenarios-creating.adoc:
```diff
@@ -0,0 +1,261 @@
+
+With this data, the test scenario can validate the expected and actual results for a decision instance based on the defined parameters.
+
+.Prerequisites
```
@dupliaka `--enable-proposed-api kiegroup.vscode-extension-pack-kogito-kie-editors` parameter is still required to launch the editors.

sterobin replied:
@dupliaka , @yesamer ah, I just confirmed that myself as well. Mine opens that way by default (bash shortcut) and we instruct users to do this already in the section about [enabling the VSCode extension](https://docs.jboss.org/kogito/release/latest/html_single/#proc-kogito-vscode-extension_kogito-creating-running). 

So technically this is already covered by their doing that prerequisite of enabling the extension, but for clarity and emphasis, I've add the following additional prerequisite to this section and to the DMN and BPMN sections in the docs.

Good with you?

```
* You have launched VSCode from a command terminal with the following command and parameters. These parameters are required to enable {PRODUCT} modelers in VSCode for DMN, BPMN, and test scenario files.
+
--
.On Linux or Windows
[source]
----
$ code --enable-proposed-api kiegroup.vscode-extension-pack-kogito-kie-editors
----

.On Mac
[source]
----
$ code --args --enable-proposed-api kiegroup.vscode-extension-pack-kogito-kie-editors
----
--
```

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/test-scenarios/proc-test-scenarios-creating.adoc:
```diff
@@ -0,0 +1,261 @@
+. In row 1, create a test scenario with values for the *GIVEN* and *EXPECT* definitions to test the decision logic of the corresponding decision service.
++
+--
+For this example, the test scenario specifies the following values:
```
Sure, but I've intentionally gone through setting up the scenarios and show the screenshot of the completed scenario file setup after the next couple steps. This is also consistent with the approach I've used in the DMN and BPMN procedures.

dupliaka replied:
Ok

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/test-scenarios/con-test-scenarios.adoc:
```diff
@@ -0,0 +1,16 @@
+When you run the test scenarios in your {PRODUCT} project, all test scenarios in all `.scesim` files are run as part of the same command. However, individual test scenarios are independent, so one scenario cannot affect or modify another. You can run test scenarios at any time during project development. You do not need to compile or deploy your decision service to run test scenarios, but your {P ...
+
+//@comment: Excluding the following until DRL-based scenarios are supported. (Stetson, 8 May 2020)
+//All data objects related to a test scenario must be in the same project package as the test scenario. Assets in the same package are imported by default.
```
Sure, changed to "Data objects", though this is currently commented out anyway since applicable to DRL-based scenarios.

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/creating-running/proc-kogito-testing-decision-logic.adoc:
```diff
@@ -0,0 +1,477 @@
+** *isAdult* -> *value*: `true`
+
++
+This example scenario tests whether the person John Quark with 20 years of age is correctly identified by the `isAdult` DMN decision as an adult (`true`), based on the decision logic that adults are more than 18 years old.
```
Good point. I've reworded as follows, and it's then re-clarified and illustrated in the concluding screenshots later.

Work for you?

> The corresponding `PersonDecisions.dmn` model in this example contains an `isAdult` DMN decision that determines that persons with more than 18 years of age are adults. This example test scenario tests whether the person John Quark with 20 years of age is correctly identified by the `isAdult` DMN decision as an adult (`true`), based on the decision logic.

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/test-scenarios/proc-test-scenarios-creating.adoc:
```diff
@@ -0,0 +1,261 @@
+.Example test scenario header cells
+image::kogito/creating-running/kogito-test-scenario-example-person-headers-create.png[Image of test scenario template]
+
+If you plan to use the same value for a *GIVEN* input condition in all test scenario rows in the file, you can use the *Background* tab to define the global header cell and value definitions and exclude the column from the test scenario table template. The *GIVEN* input conditions in the *Background* page are treated as background data for the table and are applied to all test scenarios listed.
```
@dupliaka , yeah, I went back and forth with this. Pros and cons. But good point. I've moved down to be after the completed example, before the more advanced example and final run step. Wdyt? http://file.rdu.redhat.com/~sterobin/BXMSDOC-5869_DS/#proc-test-scenarios-creating_test-scenarios

sterobin commented:
@dupliaka Done with your feedback and updated the linked doc previews. Also fixed the broken data index repo link. They apparently moved this to kogito-apps/data-index: https://github.com/kiegroup/kogito-apps/tree/master/data-index

Good with you now?

sterobin commented:
@kaldesai , all other reviews complete. Down to you. Thanks.

kaldesai commented on assemblies/assembly-kogito-developing-decision-services/main.adoc:
```diff
@@ -119,6 +119,19 @@ include::{kogito-dir}/decision-tables/ref-decision-tables-rule-table-entries.ado
+
+As a developer of business decisions, you can use test scenarios in {PRODUCT} to test the decision logic in your {PRODUCT} decision services. Test scenarios enable you to ensure that your {PRODUCT} decision services are functioning properly and as expected as you develop your {PRODUCT} project.
+
+IMPORTANT: Test scenarios in {PRODUCT} currently support DMN decision services only. Test scenarios will support Drools Rule Language (DRL) decision services in a future release.
```
Just a question: 
If you have already included this (IMPORTANT:) in the introduction part then is it required to mention it again in the next chapters?

sterobin replied:
Thanks for the review @kaldesai . I agree that it seems repetitive, but from a single-source perspective, we can't be sure that the user will have entered into this content from the intro, or even the assembly. It's a critical note so we just need to ensure that the user sees it whether starting from the intros or diving right in to the procedure. Same if we reuse any of the sections in another story, the note needs to be clear.

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/chap-kogito-release-notes.adoc:
```diff
@@ -4,7 +4,13 @@
+
+* https://issues.redhat.com/secure/ReleaseNote.jspa?projectId=12322421&version=12345352[{PRODUCT} {COMMUNITY_VERSION}]
+** https://issues.redhat.com/secure/ReleaseNote.jspa?projectId=12322421&version=12345454[{PRODUCT} tooling 0.4.1]
+** https://issues.redhat.com/secure/ReleaseNote.jspa?projectId=12322421&version=12345578[{PRODUCT} tooling 0.4.2]
```
@ederign , this is the idea ^^ Wdyt?
